dbms/src/Flash/Pipeline/Schedule/WaitQueue.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Pipeline/Schedule/WaitQueue.h>
+#include <assert.h>
+#include <common/likely.h>
+
+namespace DB
+{
+bool WaitQueue::take(std::list<TaskPtr> & local_waiting_tasks)
```

Comment:
The processing with `local_waiting_tasks`, i.e., wait or not by checking if it's empty, introduces unnecessary complexity for this `take` method. Maybe we can refine that to a more simple and natural way.

Reply:
refined to
```
    /// return false if the wait queue had been closed.
    // this function will wait until `!waiting_tasks.empty()`
    bool take(std::list<TaskPtr> & local_waiting_tasks);
    // this function will return immediately.
    bool tryTake(std::list<TaskPtr> & local_waiting_tasks);

    // Get the incremental tasks from wait queue.
    // return false if wait_queue has been closed.
    auto take_from_wait_queue = [&]() {
        return local_waiting_tasks.empty()
            ? wait_queue.take(local_waiting_tasks)
            // If the local waiting tasks are not empty, there is no need to be blocked here
            // and we can continue to process the leftover tasks in the local waiting tasks
            : wait_queue.tryTake(local_waiting_tasks);
    };
```